rmw_zenoh_cpp/src/rmw_client.cpp

```diff
@@ -0,0 +1,487 @@
+    node,
+    node->implementation_identifier,
+    eclipse_zenoh_identifier,
+    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION
+  );
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(client, RMW_RET_INVALID_ARGUMENT);
+  RMW_CHECK_ARGUMENT_FOR_NULL(result, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_FOR_NULL_WITH_MSG(
+      client->data, "client implementation pointer is null", RMW_RET_INVALID_ARGUMENT);
+
+  // OBTAIN CLIENT MEMBERS =====================================================
+  auto client_data = static_cast<rmw_client_data_t *>(client->data);
+
+  // CHECK SERVER AVAILABILITY =================================================
+  // Check if server is alive by querying its availability Zenoh queryable
+  zn_query(client_data->zn_session_,
+           client->service_name,
+           "",  // NOTE(CH3): Maybe use this predicate we want to more things in the queryable
+           zn_query_target_default(),
+           zn_query_consolidation_default(),
+           rmw_client_data_t::zn_service_availability_query_callback);
+
+  std::string key(client->service_name);
+
+  if (rmw_client_data_t::zn_availability_query_responses_.find(key)
+      != rmw_client_data_t::zn_availability_query_responses_.end()) {
+    rmw_client_data_t::zn_availability_query_responses_.erase(key);
+    *result = true;
+  }
+
+  return RMW_RET_OK;
+}
+
+// CREATE SERVICE CLIENT =======================================================
+// Create and return an rmw service client
+rmw_client_t *
+rmw_create_client(
+  const rmw_node_t * node,
+  const rosidl_service_type_support_t * type_supports,
+  const char * service_name,
+  const rmw_qos_profile_t * qos_profile)
+{
+  RCUTILS_LOG_INFO_NAMED("rmw_zenoh_cpp", "[rmw_create_client] %s", service_name);
+
+  // ASSERTIONS ================================================================
+  RMW_CHECK_ARGUMENT_FOR_NULL(node, nullptr);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(node,
+                                   node->implementation_identifier,
+                                   eclipse_zenoh_identifier,
+                                   return nullptr);
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(service_name, nullptr);
+  if (strlen(service_name) == 0) {
+    RMW_SET_ERROR_MSG("service name is empty string");
+    return nullptr;
+  }
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(qos_profile, nullptr);
+
+  // TODO(CH3): When we figure out how to spoof QoS, check for a 'configured' QoS to pass the final
+  // test that is failing
+  //
+  // if (# SOME CHECK HERE) {
+  //   RMW_SET_ERROR_MSG("expected configured QoS profile");
+  //   return nullptr;
+  // }
+
+  RMW_CHECK_ARGUMENT_FOR_NULL(type_supports, nullptr);
+
+  // OBTAIN ALLOCATOR ==========================================================
+  rcutils_allocator_t * allocator = &node->context->options.allocator;
+
+  // VALIDATE SERVICE NAME =====================================================
+  int * validation_result = static_cast<int *>(allocator->allocate(sizeof(int), allocator->state));
```

Comment:
Why are you allocating an `int` on the heap for this? (This occurs in multiple files.)

Reply:
Actually I'm not too sure.

I think I got too caught up with allocating pointers I forgot I could just get pointers on the stack :X

I'll go fix that

Reply:
Updated: 5b6237fe8d4fed3d98dad45ff8884a6931f91543